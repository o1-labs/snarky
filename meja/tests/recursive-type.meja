/* The series of unifications below will create a recursive type
   'b = (_, 'b)
   This should be handled gracefully by the typechecker.

   The staged unification
     'b = ('x, 'y) = ('z, 'b) = ('z, ('x, 'y))
   must be stopped before entering an infinite co-recursion on
     ('x, 'y) = ('z, ('x, 'y))
*/
let f = fun (f : 'a -> 'a -> 'a -> 'a, i : 'b, j : (_, _), k : (_, 'b)) => {
  f(i, j, k);
};
