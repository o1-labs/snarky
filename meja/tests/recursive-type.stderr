File "tests/recursive-type.meja", line 11, characters 10-11:
Error: Incompatible types ('c, 'd) and ('c, ('c, 'd)):
       The variable 'd would have an instance that contains itself.
